packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -3,50 +3,197 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:ui';
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/painting.dart';
+import 'package:flutter/services.dart';
 
-/// A leaf node in the focus tree that can receive focus.
+import 'binding.dart';
+import 'focus_traversal.dart';
+import 'framework.dart';
+
+/// Signature of a callback used by [Focus.onKey] and [FocusScope.onKey]
+/// to receive key events.
+///
+/// The [node] is the node that received the event.
+typedef FocusOnKeyCallback = bool Function(FocusNode node, RawKeyEvent event);
+
+/// An object that can be used by a stateful widget to obtain the keyboard focus
+/// and to handle keyboard events.
+///
+/// {@template flutter.widgets.focus_manager.focus.advancedWarning}
+/// **If you are not implementing a widget which handles focused input directly
+/// (e.g. implementing your own text field or other widget primitive), or
+/// keeping [FocusNode] references so that you can control the focus from
+/// another part of the application, you typically will not need to manage a
+/// [FocusNode] directly, as it will be handled by the [Focus] and [FocusScope]
+/// widgets or by the widget primitives themselves. Please see those widgets for
+/// typical usage.**
+/// {@endtemplate}
```

Comment:
I wouldn't make this bold. especially if you're inserting this into multiple places.

I don't know that I'd have the warning at all, though, to be honest. I mean, it's pretty common to keep FocusNode references to control the focus.

The problem with warnings like this is that people start thinking they should avoid the feature even when it's the right thing to do, and they end up doing crazy workarounds to avoid the API.

Reply:
I wouldn't worry too much about people doing things the hard way. There's only so much we can do to smooth the path.

Links to the widgets seem reasonable. How exactly you frame it is an interesting question which I'm sure I'll have opinions on...